1339 - Enabling Notes for fosterers

sarvaiyanidhi opened the pull request:
# 🔗 Issue
#1339 

# ✍️ Description
Implemented below listed changes for the functionality
*  Added new Note Model
*  Added new Notes Controller
*  Move notes modal path outside pet into shared folder
*  Remove notes from adopter_applications
*  Added Policy for notes
*  Updated notes functionality for foster and adopter application with polymorphic association

# 📷 Screenshots/Demos

![fosterer_notes](https://github.com/user-attachments/assets/6517d462-2780-4cd0-8d8e-ecf115fc4f52)

sarvaiyanidhi commented on app/controllers/organizations/staff/notes_controller.rb:
```diff
@@ -0,0 +1,43 @@
+        notable_id = params[:notable_id]
+
+        if notable_type.present? && notable_id.present?
+          klass = notable_type.constantize
```
Have updated code to check explicit types to avoid this issue

jmilljr24 commented on app/helpers/organizations/staff/notes_helper.rb:
You can delete this file.

jmilljr24 commented:
Also from the issue, can you please add the updated_at to the modal?

```
If the note already exists, should also show when the note was last modified in the modal e.g., Last modified: 04-5-25 using the updated_at value on the note record.
```

Just an idea for how it could look. Do what you think looks best. Maybe a lighter gray color from our theme.
![updated_at](https://github.com/user-attachments/assets/ff09f26d-106c-4c2d-b817-8cd822265be5)

sarvaiyanidhi commented:
Thank you @jmilljr24 for the feedback. I have noted all points and will push changes soon.

kasugaijin commented:
How's this going @sarvaiyanidhi ? 
I have been hectic lately so stepped away for a bit. Back now, and just seeing if you still want to finish this off? No rush at all.

sarvaiyanidhi commented:
> How's this going @sarvaiyanidhi ? I have been hectic lately so stepped away for a bit. Back now, and just seeing if you still want to finish this off? No rush at all.

Hi @kasugaijin I am still committed on completing this issue.. Have already fixed few of the pointed changes.. Will try to see that all changes are pushed and closed in this week. Thanks

sarvaiyanidhi commented:
Hey @kasugaijin Was bit tied up with something.. I am back now and will push changes soon. Will try if possible in couple of days.

kasugaijin commented:
@sarvaiyanidhi thanks for the update! Take your time as needed :)

sarvaiyanidhi commented on db/migrate/20250220220629_create_notes.rb:
```diff
@@ -0,0 +1,10 @@
+class CreateNotes < ActiveRecord::Migration[8.0]
+  def change
+    create_table :notes do |t|
+      t.text :notes
```
@jmilljr24 How about changing attribute name to 'content' instead of 'notes' as I think body might be confusing.. Can you confirm if content is fine or you want to go ahead with body?

jmilljr24 replied:
@sarvaiyanidhi go with "content" Thanks!

sarvaiyanidhi commented on app/controllers/organizations/staff/notes_controller.rb:
```diff
@@ -0,0 +1,51 @@
+        @context = params[:context] || "default"
+
+        respond_to do |format|
+          format.html { redirect_back(fallback_location: root_path, notice: t("organizations.staff.notes.update.success")) }
```
Have updated code to update deprecated syntax.

sarvaiyanidhi commented on app/controllers/organizations/staff/notes_controller.rb:
```diff
@@ -0,0 +1,51 @@
+      end
+
+      def note_params
+        params.permit(:notes, :notable_id, :notable_type)
```
Fixed with new rails 8 syntax of params.expect which resolved unpermitted param error in log.

sarvaiyanidhi commented on app/views/organizations/staff/shared/_notes.html.erb:
```diff
@@ -0,0 +1,42 @@
+        <button type="button" class="btn-close" data-bs-dismiss="modal" aria-label="Close"></button>
+      </div>
+      <div class="modal-body border-top-0 pt-2">
+        <%= bootstrap_form_with url: staff_note_path, method: "patch", data: { turbo: true } do |form| %>
```
Implemented the changes

sarvaiyanidhi commented on test/policies/organizations/note_policy_spec.rb:
```diff
@@ -0,0 +1,79 @@
+require "rails_helper"
+
+RSpec.describe Organizations::NotePolicy, type: :policy do
```
Sorry it was a mistake. Not sure how it happened but may be as I was looking into some other code which is using Rspec. Have made the changes to use minitest.

sarvaiyanidhi commented on app/views/organizations/staff/notes/update.turbo_stream.erb:
```diff
@@ -0,0 +1,17 @@
+<%= turbo_stream.replace dom_id(@notable, "notes_display_#{@context}") do %>
```
Have made the necessary changes to use replace_all and tested functionality using it.

sarvaiyanidhi commented on db/seeds/01_alta.rb:
Have added seed data for new structure.

sarvaiyanidhi commented:
@kasugaijin @jmilljr24 Really sorry for the delay in getting required changes done. I have completed all addressed points and pushed the code. Can you please review and let me know if there are any more changes to be done. I am available now and will address any more changes quickly to close this issue soon.

jmilljr24 reviewed (approved):
@sarvaiyanidhi Looks great! We appreciate your work on this!!!

kasugaijin commented:
Nice work @sarvaiyanidhi
